[AOTI][Intel GPU] Support multi_arch_kernel_binary option for XPU.

etaf opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #154514

Following the design of #154413, this PR add XPU support for generating kernel binary files that support multiple archs.

Fixes #154682, Fixes #154683, Fixes 154689, Fixes #154685 , Fixes #154690, Fixes #154681

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

etaf committed:
[AOTI][Intel GPU] Support multi_arch_kernel_binary option for XPU.

[ghstack-poisoned]

desertfire commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -1003,7 +1003,9 @@ def save_gpu_kernel(self, stream, launcher):
         binary = launcher.bin.asm[bin_type]
         # Also store asm code which can be used for debugging and generating cpp package
-        asm_type = {"hip": "amdgcn", "cuda": "ptx"}.get(self.device_props.type, None)
+        asm_type = {"hip": "amdgcn", "cuda": "ptx", "xpu": "spv"}.get(
```
Since we already generate spv for xpu, shouldn't the added new option be a no-op for xpu?

etaf replied:
Hi @desertfire , Yes, in theory XPU doesn’t need this option. I added it to align with the user interface, and also to fix some errors that occur on XPU when this option is specified in UT. For example:  https://hud.pytorch.org/pr/pytorch/pytorch/147693#43085076369
```
RuntimeError: Failed to run autotuning code block: Missing kernel assembly code
```

And this may also helps avoid adding multiple if-else checks in the code specifically for XPU.

desertfire reviewed (commented):
I also did some config renaming in https://github.com/pytorch/pytorch/pull/154608. You will need to rebase after that.

desertfire commented on torch/_inductor/codecache.py:
```diff
@@ -407,7 +407,7 @@ def get_path(
-    if hash_type in {"amdgcn", "code", "ptx"}:
+    if hash_type in {"amdgcn", "code", "ptx", "spv"}:
         return code_hash(content, extra)
     if hash_type in {"cubin", "hsaco", "spv"}:
```
It probably doesn't matter in terms of functionality, but "spv" will not hit this branch anymore.

etaf committed:
Update on "[AOTI][Intel GPU] Support multi_arch_kernel_binary option for XPU."


Following the design of #154413, this PR add XPU support for generating kernel binary files that support multiple archs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

etaf committed:
Update on "[AOTI][Intel GPU] Support multi_arch_kernel_binary option for XPU."


Following the design of #154413, this PR add XPU support for generating kernel binary files that support multiple archs.

Fixes #154682, Fixes #154683, Fixes 154689, Fixes #154685 , Fixes #154690, Fixes #154681

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

etaf commented:
@pytorchbot merge
